File: samples/ReactTypeScriptApp/AppHost/src/app-host.ts
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { cloudshell } from "../../../../Launchers/TypeScript/cloudshell/src/index.ts";

const appHostRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const sampleRoot = resolve(appHostRoot, "..");
const repoRoot = resolve(sampleRoot, "..", "..");
const frontendRoot = resolve(sampleRoot, "App");
const backendRoot = resolve(sampleRoot, "Backend");
const cliProject = pathFromEnv(
  "CLOUDSHELL_CLI_PROJECT",
  resolve(repoRoot, "CloudShell.Cli", "CloudShell.Cli.csproj"));
const hostProject = pathFromEnv(
  "CLOUDSHELL_HOST_PROJECT",
  resolve(repoRoot, "CloudShell.LocalDevelopmentHost", "CloudShell.LocalDevelopmentHost.csproj"));
const stateDir = pathFromEnv("CLOUDSHELL_STATE_DIR", resolve(sampleRoot, ".cloudshell"));
const dataDir = readArgumentValue("--data-dir") ??
  pathFromEnv("CLOUDSHELL_DATA_DIR", stateDir);
const controlPlaneUrl = pathFromEnv(
  "CLOUDSHELL_CONTROL_PLANE_URL",
  "http://127.0.0.1:5110");
const frontendEndpoint = endpointFromEnv(
  "ReactTypeScriptApp__FrontendEndpoint",
  "http://localhost:5175");
const backendEndpoint = endpointFromEnv(
  "ReactTypeScriptApp__BackendEndpoint",
  "http://localhost:5185");
const settingsServiceEndpoint = pathFromEnv(
  "ReactTypeScriptApp__ConfigurationServiceEndpoint",
  "http://localhost:5111");
const edgePort = numberFromEnv("ReactTypeScriptApp__EdgeHttpPort", 8088);
const settingsResourceId = "configuration.store:react-typescript-settings";
const settingsApiEndpoint =
  `${settingsServiceEndpoint.replace(/\/$/, "")}/api/configuration/stores/${encodeURIComponent(settingsResourceId)}/settings`;

const app = cloudshell("react-typescript-app", {
  metadata: {
    "cloudshell.source": "typescript",
    "cloudshell.sample": "ReactTypeScriptApp"
  }
});

const settings = app
  .addConfigurationStore("react-typescript-settings")
  .withDisplayName("React TypeScript Settings")
  .withEndpoint(settingsServiceEndpoint)
  .withSeed(seed => seed
    .setting("Sample--Message", "Hello from the CloudShell Configuration Store")
    .setting("Sample--Mode", "react-with-backend"));

const backend = app
  .addJavaScriptApp("react-api", backendRoot)
  .withDisplayName("React Sample Backend")
  .withPackageManager("npm")
  .withScript("dev")
  .withServiceDiscovery()
  .withReference(settings)
  .dependsOn(settings)
  .withEnvironmentVariable("PORT", backendEndpoint.port.toString())
  .withEnvironmentVariable("CLOUDSHELL_SETTINGS_ENDPOINT", settingsApiEndpoint)
  .withEnvironmentVariable("OTEL_SERVICE_NAME", "react-sample-api")
  .withEnvironmentVariable("Sample__Message", {
    configurationSettingRef: settings.setting("Sample--Message")
  })
  .withEnvironmentVariable("Sample__Mode", {
    configurationSettingRef: settings.setting("Sample--Mode")
  })
  .withHttpEndpoint({
    host: backendEndpoint.hostname,
    port: numberFromEndpoint(backendEndpoint),
    targetPort: numberFromEndpoint(backendEndpoint)
  })
  .withHttpHealthCheck("/healthz", { endpointName: "http" })
  .withHttpLivenessCheck("/alive", { endpointName: "http" });

const frontend = app
  .addJavaScriptApp("react-frontend", frontendRoot)
  .withDisplayName("React TypeScript Frontend")
  .withPackageManager("npm")
  .withScript("dev")
  .withServiceDiscovery()
  .withReference(backend)
  .dependsOn(backend)
  .withEnvironmentVariable("PORT", frontendEndpoint.port.toString())
  .withEnvironmentVariable("VITE_BACKEND_URL", backendEndpoint.toString().replace(/\/$/, ""))
  .withEnvironmentVariable("VITE_SAMPLE_TITLE", "CloudShell React TypeScript")
  .withHttpEndpoint({
    host: frontendEndpoint.hostname,
    port: numberFromEndpoint(frontendEndpoint),
    targetPort: numberFromEndpoint(frontendEndpoint)
  })
  .withHttpHealthCheck("/", { endpointName: "http" });

app
  .addLoadBalancer("react-edge")
  .withDisplayName("React Sample Edge")
  .withProvider("traefik")
  .exposeHttp({ port: edgePort })
  .mapPath("react.localhost", "/api", backend, { endpoint: "http" })
  .mapHost("react.localhost", frontend, { endpoint: "http" });

if (process.argv.includes("--apply") || process.argv.includes("--start") || process.argv.includes("--run")) {
  const options = {
    cliProject,
    hostProject,
    noBuild: process.argv.includes("--no-build"),
    controlPlaneUrl,
    url: controlPlaneUrl,
    stateDir,
    dataDir,
    cwd: repoRoot
  };
  const result = process.argv.includes("--run")
    ? await app.run(options)
    : process.argv.includes("--start")
      ? await app.start(options)
      : await app.apply(options);

  process.exitCode = result.exitCode;
} else {
  process.stdout.write(app.toJson());
}

function pathFromEnv(name: string, fallback: string): string {
  return process.env[name] && process.env[name]!.trim().length > 0
    ? process.env[name]!
    : fallback;
}

function endpointFromEnv(name: string, fallback: string): URL {
  return new URL(pathFromEnv(name, fallback));
}

function numberFromEnv(name: string, fallback: number): number {
  const value = process.env[name];
  if (!value || value.trim().length === 0) {
    return fallback;
  }

  return Number.parseInt(value, 10);
}

function numberFromEndpoint(endpoint: URL): number {
  return Number.parseInt(endpoint.port, 10);
}

function readArgumentValue(name: string): string | undefined {
  const index = process.argv.findIndex(argument =>
    argument.toLowerCase() === name.toLowerCase());
  return index >= 0 ? process.argv[index + 1] : undefined;
}

File: samples/JavaContainerApp/AppHost/AppHost.java
import com.cloudshell.launcher.CloudShellApp;
import com.cloudshell.launcher.CloudShellLauncherOptions;
import com.cloudshell.launcher.ConfigurationStoreResource;
import com.cloudshell.launcher.NetworkResource;
import com.cloudshell.launcher.SecretsVaultResource;
import java.nio.file.Files;
import java.nio.file.Path;

public final class AppHost {
    private AppHost() {
    }

    public static void main(String[] args) throws Exception {
        Path repoRoot = findRepositoryRoot(Path.of("").toAbsolutePath());
        Path sampleRoot = repoRoot.resolve("samples").resolve("JavaContainerApp");
        Path javaAppRoot = sampleRoot.resolve("App");
        Path cliProject = pathFromEnv(
                "CLOUDSHELL_CLI_PROJECT",
                repoRoot.resolve("CloudShell.Cli").resolve("CloudShell.Cli.csproj"));
        Path hostProject = pathFromEnv(
                "CLOUDSHELL_HOST_PROJECT",
                repoRoot.resolve("CloudShell.LocalDevelopmentHost").resolve("CloudShell.LocalDevelopmentHost.csproj"));
        Path stateDir = pathFromEnv(
                "CLOUDSHELL_STATE_DIR",
                sampleRoot.resolve(".cloudshell"));
        Path dataDir = Path.of(argumentValue(args, "--data-dir",
                System.getenv().getOrDefault("CLOUDSHELL_DATA_DIR", stateDir.toString())));
        String controlPlaneUrl = argumentValue(args, "--control-plane",
                System.getenv().getOrDefault("CLOUDSHELL_CONTROL_PLANE_URL", "http://127.0.0.1:5112"));

        CloudShellApp app = buildTemplate(repoRoot, javaAppRoot);

        if (hasArgument(args, "--apply") || hasArgument(args, "--start") || hasArgument(args, "--run")) {
            CloudShellLauncherOptions options = new CloudShellLauncherOptions()
                    .withCliProject(cliProject)
                    .withHostProject(hostProject)
                    .withControlPlaneUrl(controlPlaneUrl)
                    .withHostUrl(controlPlaneUrl)
                    .withStateDirectory(stateDir)
                    .withDataDirectory(dataDir)
                    .withTemplatePath(argumentPath(args, "--template-path", stateDir.resolve("resources.json")))
                    .withWorkingDirectory(repoRoot)
                    .withNoBuild(hasArgument(args, "--no-build"));

            if (hasArgument(args, "--run")) {
                System.exit(app.run(options).exitCode());
            }

            if (hasArgument(args, "--start")) {
                System.exit(app.start(options).exitCode());
            }

            System.exit(app.apply(options).exitCode());
        }

        System.out.print(app.toJson());
    }

    private static CloudShellApp buildTemplate(Path repoRoot, Path javaAppRoot) {
        CloudShellApp app = CloudShellApp.create("java-container-app")
                .withEnvironmentId("local")
                .withMetadata("cloudshell.source", "java")
                .withMetadata("cloudshell.sample", "JavaContainerApp");

        NetworkResource hostNetwork = app.addNetwork("host")
                .withResourceId("network:host")
                .withDisplayName("Host network")
                .withNetworkKind("Host")
                .withHostReadiness("hostReady");

        ConfigurationStoreResource settings = app.addConfigurationStore("java-container-settings")
                .withDisplayName("Java Container Settings")
                .withEndpoint("http://localhost:5113")
                .withSeed(seed -> seed.setting("Sample--Message", "Hello from Java container app configuration"));

        SecretsVaultResource secrets = app.addSecretsVault("java-container-secrets")
                .withDisplayName("Java Container Secrets")
                .withEndpoint("http://localhost:6113")
                .withSeed(seed -> seed.secret("Sample--ApiKey", "java-container-secret", "v1"));

        var api = app.addJavaApp(
                "java-container-api",
                javaAppRoot.toString(),
                "target/cloudshell-java-container-app-sample.jar")
                .withDisplayName("Java Container API")
                .withServiceDiscovery()
                .withEnvironmentVariable("PORT", "8080")
                .withEnvironmentVariable("OTEL_SERVICE_NAME", "java-container-api")
                .withEnvironmentVariable("CLOUDSHELL_CONFIGURATION_SERVICE_NAME", "java-container-settings")
                .withEnvironmentVariable("CLOUDSHELL_SECRETS_VAULT_NAME", "java-container-secrets")
                .withReference(settings)
                .withReference(secrets)
                .dependsOn(settings)
                .dependsOn(secrets)
                .withHttpEndpoint("localhost", 5191, 8080, hostNetwork)
                .withHttpHealthCheck("/healthz")
                .withHttpLivenessCheck("/alive")
                .withDefaultConsoleLogSource()
                .asContainerApp(
                    null,
                    null,
                    "dev",
                    repoRoot.toString(),
                    "samples/JavaContainerApp/App/Dockerfile",
                    2)
                .requireIdentity("java-container-api")
                .provisionIdentityOnStartup();

        settings.allowResourceIdentity(
            api,
            "CloudShell.Configuration/stores/settings/read/action",
            "java-container-api");
        secrets.allowResourceIdentity(
            api,
            "CloudShell.Secrets/vaults/secrets/read/action",
            "java-container-api");

        return app;
    }

    private static Path pathFromEnv(String name, Path fallback) {
        String value = System.getenv(name);
        return value == null || value.isBlank() ? fallback : Path.of(value);
    }

    private static Path argumentPath(String[] args, String name, Path fallback) {
        String value = argumentValue(args, name, null);
        return value == null || value.isBlank() ? fallback : Path.of(value);
    }

    private static boolean hasArgument(String[] args, String name) {
        for (String arg : args) {
            if (arg.equalsIgnoreCase(name)) {
                return true;
            }
        }

        return false;
    }

    private static String argumentValue(String[] args, String name, String fallback) {
        for (int index = 0; index < args.length - 1; index++) {
            if (args[index].equalsIgnoreCase(name)) {
                return args[index + 1];
            }
        }

        return fallback;
    }

    private static Path findRepositoryRoot(Path start) {
        Path directory = start;
        while (directory != null) {
            if (Files.exists(directory.resolve("CloudShell.slnx"))) {
                return directory;
            }

            directory = directory.getParent();
        }

        throw new IllegalStateException("Could not find CloudShell.slnx from " + start);
    }
}
